DatadogCore/Sources/Core/Upload/DataUploadWorker.swift

```diff
@@ -55,95 +63,113 @@ internal class DataUploadWorker: DataUploadWorkerType {
+                        )
                     }
 
-                    switch uploadStatus.error {
-                    case .unauthorized:
-                        DD.logger.error("⚠️ Make sure that the provided token still exists and you're targeting the relevant Datadog site.")
-                    case let .httpError(statusCode: statusCode):
-                        telemetry.error("Data upload finished with status code: \(statusCode)")
-                    case let .networkError(error: error):
-                        telemetry.error("Data upload finished with error", error: error)
-                    case .none: break
+                    if let error = uploadStatus.error {
+                        switch error {
+                        case .unauthorized:
+                            DD.logger.error("⚠️ Make sure that the provided token still exists and you're targeting the relevant Datadog site.")
+                        case let .httpError(statusCode: statusCode):
+                            telemetry.error("Data upload finished with status code: \(statusCode)")
+                        case let .networkError(error: error):
+                            telemetry.error("Data upload finished with error", error: error)
+                        }
                     }
                 } catch let error {
                     // If upload can't be initiated do not retry, so drop the batch:
                     self.fileReader.markBatchAsRead(batch, reason: .invalid)
                     telemetry.error("Failed to initiate '\(self.featureName)' data upload", error: error)
                 }
+            }
+            if files.isEmpty {
+                self.scheduleNextCycle()
             } else {
-                let batchLabel = nextBatch != nil ? "YES" : (isSystemReady ? "NO" : "NOT CHECKED")
-                DD.logger.debug("💡 (\(self.featureName)) No upload. Batch to upload: \(batchLabel), System conditions: \(blockersForUpload.description)")
-
-                self.delay.increase()
-                self.backgroundTaskCoordinator?.endBackgroundTask()
+                self.uploadFile(from: files, context: context)
             }
-
-            self.scheduleNextUpload(after: self.delay.current)
         }
-
         self.uploadWork = uploadWork
-
-        scheduleNextUpload(after: self.delay.current)
-    }
-
-    private func scheduleNextUpload(after delay: TimeInterval) {
-        guard let work = uploadWork else {
-            return
-        }
-
-        queue.asyncAfter(deadline: .now() + delay, execute: work)
+        queue.async(execute: uploadWork)
     }
 
     /// Sends all unsent data synchronously.
     /// - It performs arbitrary upload (without checking upload condition and without re-transmitting failed uploads).
     internal func flushSynchronously() {
-        queue.sync {
-            while let nextBatch = self.fileReader.readNextBatch() {
+        queue.sync { [fileReader, dataUploader, contextProvider] in
```

Comment:
is capturing of these params are required?

I wonder what happens this class is de-allocated and there is still some files are pending.

Reply:
I can probably can fallback to just `[weak self]`.

This object would only get deallocated when we destroy core.